### PR TITLE
Add skill evaluation receipt aggregation and grading

### PR DESCRIPTION
**Pull request by howarewoo**
## Goal

Turn prepared skill-evaluation runs into trustworthy aggregate evidence without adding the report-rendering surface yet.

## Summary

- Validate exact manifests and evidence with safe file access, immutable snapshots, and frozen package/definition proof.
- Evaluate deterministic assertions and blinded paired qualitative grades, then calculate trigger metrics and complete, blocked, or degraded outcomes.
- Publish aggregate JSON through durable create-new semantics with focused integrity, grading, metric, and failure-path coverage.

### Automated

- [ ] `bash skills/woostack-eval/scripts/tests/test-aggregate.sh` — PASS

### Manual

**Before merge**

- [ ] Inspect blocked and degraded fixtures to confirm incomplete or mismatched proof never becomes a clean comparison.
- [ ] Confirm qualitative grades remain blind until candidate and baseline mappings plus grader receipts validate.
- [ ] Confirm existing aggregate outputs are never overwritten and publication failures leave no partial final artifact.

Spec: .woostack/specs/2026-07-15-skill-evaluation-optimization.md

### Automated

- [ ] `bash skills/woostack-eval/scripts/tests/run-tests.sh` — passed all 9 evaluator script suites.
- [ ] `bash skills/woostack-eval/scripts/tests/test-critical-corpora.sh` — validated exactly 15 required packages.
- [ ] `bash skills/woostack-review/scripts/tests/test-prefetch-skill-package.sh` — passed 45/45.
- [ ] `bash skills/woostack-review/scripts/tests/test-prefetch-skill-package-ci.sh` — passed 18/18.
- [ ] `bash skills/woostack-review/scripts/tests/test-prefetch-skill-small-diff.sh` — passed 19/19.
- [ ] `bash skills/woostack-review/scripts/tests/test-skills-angle-rubric.sh` — passed 28/28.
- [ ] `bash skills/woostack-review/scripts/tests/test-progressive-disclosure.sh` — passed 190/190.
- [ ] `bash skills/woostack-commit/tests/test-linear-attribution.sh` — passed.
- [ ] `bash skills/woostack-commit/tests/test-progressive-disclosure.sh` — passed.
- [ ] `bash skills/woostack-build/tests/test-linear-build-contract.sh` — passed.
- [ ] `bash skills/woostack-build/scripts/tests/test-build-spec-commit-ordering.sh` — passed.
- [ ] `bash skills/woostack-build/scripts/tests/test-progressive-disclosure.sh` — passed 63/63.
- [ ] `bash skills/using-woostack/tests/test-artifact-reader-contract.sh` — passed 12/12.
- [ ] `bash skills/woostack-change/scripts/tests/test-command-surface.sh` — passed 12/12.
- [ ] `bash skills/woostack-respond/scripts/tests/test-command-surface.sh` — passed 15/15.
- [ ] `node --test site/scripts/gen-skills.test.mjs` — passed 14/14.
- [ ] `pnpm -C site build` — passed; 147 static pages generated.

### Comparative model evidence

- [ ] No model-backed old-versus-candidate run was dispatched: this host cannot prove the evaluator's required per-worker capability isolation. Therefore no receipt or report path exists, and no model-eval pass is claimed.

### CODE REVIEW

**Comment by howarewoo** [comment]
* **#555** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/555?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#556** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/556?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#553** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/553?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#552** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/552?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#551** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/551?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#550** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/550?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#548** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/548?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#546** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/546?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#543** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/543?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#541** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/541?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#540** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#539** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#538** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/538?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/howarewoo/woostack/538?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#537** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/537?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#534** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#530** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/530?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: skills/woostack-eval/scripts/aggregate/core.mjs [hunk 1]
```diff
@@ -0,0 +1,1167 @@
...
+    if (linked.kind !== 'file') {
+      const code = linked.kind === 'missing'
+        && grade.receipt.path === `${evidencePrefix}/action.grader.${grade.caseId}.${filenameIdentity.variant}.${grade.repetition}.json`
+        ? 'missing-grade-receipt'
+        : 'non-regular-evidence';
+      errors.push(errorRecord(code, '/receipt/path', gradePath, 'Linked grader receipt is absent or not a regular file'));
+      continue;
+    }
+    const expectedGraderReceiptPath =
+      `${evidencePrefix}/action.grader.${grade.caseId}.${filenameIdentity.variant}.${grade.repetition}.json`;
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/core.mjs]: **Grader receipt path drops the grader ID**

`schemas.md` requires `action.grader.<case>.<variant>.<repetition>.<grader-id>.json`, but this branch constructs the expected path without `<grader-id>`, so every conforming qualitative grade is rejected with `identity-mismatch` and its correctly named receipt is later reported as unknown.

Fix: Include the resolved grading-plan `graderId` in every expected, referenced, missing, and linked grader-receipt path, then update the aggregate fixtures to use the canonical filename.

<sub>— <strong>HIGH · BLOCKING</strong> · <code>bugs</code></sub>

File: skills/woostack-eval/scripts/aggregate/safe-access.mjs [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+  } catch (error) {
+    if (['ENOENT', 'ENOTDIR', 'ELOOP'].includes(error?.code)) return null;
+    throw error;
+  }
+  try {
+    const before = await handle.stat({ bigint: true });
+    if (before.isDirectory()) return { kind: 'directory', identity: statIdentity(before) };
+    if (!before.isFile()) return { kind: 'special', identity: statIdentity(before) };
+    const hash = createHash('sha256');
+    for await (const chunk of handle.createReadStream({ autoClose: false })) hash.update(chunk);
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/safe-access.mjs]: **Nit: Snapshot hashing permits resource exhaustion**

`captureRunSnapshot` calls this unbounded hash loop for every worker-created file with no per-file, cumulative-byte, or entry-count budget, so a permitted shell-workspace worker can leave a huge file or tree that forces three full snapshot traversals and exhausts host CPU, I/O, or memory before aggregation.

Fix: Enforce documented per-file, cumulative-byte, and entry-count ceilings before and during snapshot capture, failing the run deterministically when any ceiling is exceeded.

<sub>— <strong>MEDIUM · NIT</strong> · <code>security</code></sub>

File: skills/woostack-eval/scripts/aggregate/core.mjs [hunk 3]
```diff
@@ -0,0 +1,1167 @@
...
+
+async function evaluateAssertion(assertion, context) {
+  const base = { assertionId: assertion.id, critical: assertion.critical === true };
+  if (assertion.kind === 'qualitative') {
+    return { ...base, pass: null, rationale: null, anonymizedOutputId: null, grade: null, graderReceipt: null };
+  }
+
+  let pass = false;
+  let observed = null;
+  if (assertion.kind === 'final-contains' || assertion.kind === 'final-excludes') {
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/core.mjs:383]: **Nit: Objective assertion branches lack coverage**

The aggregate fixtures exercise only `final-contains` and `path-exists`, leaving the new `final-excludes`, `receipt-field-equals`, `path-absent`, `file-contains`, `file-excludes`, and `json-path-equals` branches without observable-behavior coverage.

Fix: Add aggregate fixtures proving passing and failing outcomes for every unexercised objective assertion kind, including missing-file and malformed-JSON outcomes where applicable.

<sub>— <strong>MEDIUM · NIT</strong> · <code>tests</code></sub>

**Review by howarewoo — COMMENTED** [review]
## Review summary

3 validated finding(s) remained after adversarial review; see inline comments for exact evidence and fixes.

---
**Status: CHANGES REQUESTED** — 1 blocking + 0 non-blocking finding(s) (1 HIGH, 0 MEDIUM, 0 LOW) + 2 nit(s). See inline comments.
*Audited by woostack-review · Host: OMP · Provider: OpenAI · Model: openai-codex/gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks REQUEST_CHANGES on your own PR. Status line above carries the actual verdict._

File: skills/woostack-eval/scripts/aggregate/core.mjs (hunk 1 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/core.mjs]: Fixed in e6e202f.

File: skills/woostack-eval/scripts/aggregate/safe-access.mjs (hunk 2 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/safe-access.mjs]: Fixed in e6e202f.

File: skills/woostack-eval/scripts/aggregate/core.mjs (hunk 3 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/core.mjs:383]: Fixed in e6e202f.

File: skills/woostack-eval/scripts/aggregate/core.mjs [hunk 4]
```diff
@@ -0,0 +1,1185 @@
...
+      context.fileCache,
+    );
+    const exists = state.kind === 'file';
+    pass = assertion.kind === 'path-exists' ? exists : state.kind === 'missing';
+    observed = { path: assertion.path };
+  } else if (['file-contains', 'file-excludes', 'json-path-equals'].includes(assertion.kind)) {
+    const relativePath = assertion.path ?? assertion.file;
+    const state = await assertionPathState(context.workspaceRoot, relativePath, context.fileCache);
+    if (state.kind === 'file') {
+      observed = { path: relativePath, sha256: state.sha256, bytes: state.byteCount };
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/core.mjs:400]: **Nested absent paths fail their assertion**

A valid `path-absent` assertion such as `missing-dir/result.txt` evaluates false because `regularFile` classifies a missing parent directory as `non-regular`, while this branch passes only the `missing` state.

Fix: Distinguish a missing ancestor from an unsafe or non-directory ancestor in `regularFile` and return `missing` for the former so nested absent paths pass.

<sub>— <strong>MEDIUM · BLOCKING</strong> · <code>bugs</code></sub>

File: skills/woostack-eval/references/schemas.md [hunk 5]
```diff
@@ -428,6 +429,7 @@ Aggregate evidence and fatal snapshot errors use these exhaustive canonical code
 | `package-hash-mismatch` | `/packageHash` | A copied package, worker receipt, or grader receipt differs from the applicable frozen manifest package identity. |
```
Comment (howarewoo) [inline skills/woostack-eval/references/schemas.md:429]: **Nit: Snapshot limits are duplicated across references**

The new `snapshot-limit-exceeded` schema row repeats the exact 4,096-entry, 16 MiB-file, and 128 MiB-run limits already added to `runner.md`, so the two reference authorities can drift despite the explicit cross-link-only rule.

Fix: Replace the repeated thresholds with a link to the canonical snapshot-bound definition in `runner.md`.

```suggestion
| `snapshot-limit-exceeded` | empty pointer | Fatal: a [snapshot bound](runner.md#isolated-workspaces-and-evidence) is crossed; publication is refused. |
```

<sub>— <strong>LOW · NIT</strong> · <code>conventions</code></sub>

File: skills/woostack-eval/scripts/aggregate/core.mjs [hunk 6]
```diff
@@ -0,0 +1,1185 @@
...
+  requireManifest(manifest);
+  await requireManifestWorkspaces(runRoot, manifest);
+  await requireQuiescenceProof(runRoot, manifest);
+  const runSnapshot = await buildRunSnapshot(runRoot);
+  const gradingPlanByKey = await requireResolvedGradingPlan(runRoot, manifest);
+
+  const evidencePrefix = relativeTo(runRoot, evidenceRoot);
+  const names = await listDirectoryNames(evidenceRoot);
+  const evidencePaths = names.map((name) => `${evidencePrefix}/${name}`);
+  const errors = [];
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/core.mjs:527]: **Snapshot does not bind evidence reads**

The captured `runSnapshot` is used only for a final tree comparison, while manifest, receipt, grade, definition, and package reads never prove that the bytes they consumed match the captured entry; a worker-controlled background process can temporarily substitute evidence during an individual read and restore the snapshotted bytes before final revalidation, causing attacker-chosen evidence to be accepted into a trustworthy aggregate.

Fix: Bind every consumed file and directory to its captured snapshot identity and SHA-256 at read time, rejecting any mismatch before evaluating or publishing the aggregate.

<sub>— <strong>HIGH · BLOCKING</strong> · <code>security</code></sub>

File: skills/woostack-eval/scripts/tests/test-aggregate.sh [hunk 7]
```diff
@@ -0,0 +1,2869 @@
...
+);
+
+const pathDirectory = await cloneRun('one-run', 'path-directory');
+const pathDirectoryDefinitionPath =
+  path.join(pathDirectory, 'definitions', 'behavior.objective-check.json');
+const pathDirectoryDefinition =
+  JSON.parse(await readFile(pathDirectoryDefinitionPath, 'utf8'));
+pathDirectoryDefinition.assertions.push({
+  id: 'directory-is-not-regular',
+  kind: 'path-exists',
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/tests/test-aggregate.sh:1500]: **Nit: Path-exists has no successful case**

The `path-exists` assertion appears only in this directory-rejection fixture and is expected to fail, so an implementation that always returns false for this supported assertion kind satisfies every test.

Fix: Add an existing regular-file fixture that must pass `path-exists` and a missing-path fixture that must fail it.

<sub>— <strong>MEDIUM · NIT</strong> · <code>tests</code></sub>

**Review by howarewoo — COMMENTED** [review]
## Review summary

The aggregate remains vulnerable to evidence substitution during reads and mishandles nested absent-path assertions; two low-risk coverage and documentation nits also remain.

---
**Status: CHANGES REQUESTED** — 2 blocking + 0 non-blocking finding(s) (1 HIGH, 1 MEDIUM, 0 LOW) + 2 nit(s). See inline comments.
*Audited by woostack-review · Host: omp · Provider: openai-codex · Model: gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks REQUEST_CHANGES on your own PR. Status line above carries the actual verdict._

File: skills/woostack-eval/scripts/aggregate/core.mjs (hunk 4 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/core.mjs:400]: Fixed in 10001aa.

File: skills/woostack-eval/references/schemas.md (hunk 5 above)
Comment (howarewoo) [reply skills/woostack-eval/references/schemas.md:429]: Fixed in 10001aa.

File: skills/woostack-eval/scripts/aggregate/core.mjs (hunk 6 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/core.mjs:527]: Fixed in 10001aa.

File: skills/woostack-eval/scripts/tests/test-aggregate.sh (hunk 7 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/tests/test-aggregate.sh:1500]: Fixed in 10001aa.

File: skills/woostack-eval/scripts/aggregate/core.mjs [hunk 8]
```diff
@@ -0,0 +1,1195 @@
...
+  const evidenceRoot = path.resolve(options.evidence);
+  const outPath = path.resolve(options.out);
+  const runRoot = path.dirname(manifestPath);
+  if (!contained(runRoot, evidenceRoot) || evidenceRoot === runRoot) throw new Error('evidence directory must be contained by the run root');
+  if (!contained(runRoot, outPath) || outPath === runRoot) throw new Error('output must be contained by the run root');
+  await requirePrivateRunRoot(runRoot);
+
+  await requireDirectoryChain(runRoot, evidenceRoot, 'evidence directory');
+  await requireDirectoryChain(runRoot, path.dirname(outPath), 'output parent');
+  const manifestRelativePath = relativeTo(runRoot, manifestPath);
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/core.mjs:513]: **Nit: Containment guards lack regression coverage**

The aggregator adds explicit run-root containment guards for both the evidence directory and output path, but test-aggregate.sh invokes only contained paths, so a regression that permits external evidence reads or out-of-root publication would pass the suite.

**Fix:** Add focused rejection cases for an evidence directory outside the run root and an output path outside the run root, asserting failure and absence of publication.

File: skills/woostack-eval/scripts/aggregate/safe-access.mjs [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+function publicationDirectoryIdentity(stats) {
+  return {
+    dev: String(stats.dev),
+    ino: String(stats.ino),
+    uid: String(stats.uid),
+    mode: String(stats.mode),
+  };
+}
+
+async function requirePrivatePublicationParent(parent) {
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/aggregate/safe-access.mjs:488]: **Nit: Private publication parent is untested**

The exported publication authority rejects missing, public, non-owner, and symlink parents, yet its tests create only mode-0700 real directories, leaving the owner-private boundary undefended against regression.

**Fix:** Add focused writeCreateNew rejection cases for a mode-0755 parent and a symlink parent, asserting that neither a final file nor a new temporary file is created.

File: skills/woostack-eval/scripts/tests/test-aggregate.sh [hunk 10]
```diff
@@ -0,0 +1,2900 @@
...
+const runId = '20260715T120000Z-4242';
+const targetSkill = 'woostack-example';
+let packageHash = null;
+const baseline = { kind: 'git-ref', identity: '0123456789abcdef0123456789abcdef01234567' };
+const materializationLimit = 1024 * 1024;
+const runConfiguration = {
+  host: 'omp',
+  runner: 'worker',
+  model: 'model-a',
+  sessionIdentity: null,
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/tests/test-aggregate.sh:43]: **Nit: Valid session identity path has no positive test**

Every successful fixture sets `model` to `model-a` and `sessionIdentity` to null, while the session-identity mutations cover only invalid empty, wrong-type, or dual-identity forms, so the documented valid session-default completion path can regress without this suite failing.

**Fix:** Add a complete paired fixture with `model: null` and one shared valid `sessionIdentity`, then assert that aggregation succeeds and preserves the completion identity.

File: skills/woostack-eval/scripts/tests/test-aggregate.sh [hunk 11]
```diff
@@ -0,0 +1,2900 @@
...
+    runner: runConfiguration.runner,
+    model: runConfiguration.model,
+    sessionIdentity: runConfiguration.sessionIdentity,
+    tier: runConfiguration.tier,
+    effort: runConfiguration.effort,
+    startedAt: `2026-07-15T12:00:0${repetition}.000Z`,
+    durationMs: variant === 'candidate' ? repetition * 100 : repetition * 200 + 100,
+    output,
+    transcript,
+    tokenUsage: 'unavailable',
```
Comment (howarewoo) [inline skills/woostack-eval/scripts/tests/test-aggregate.sh:185]: **Nit: Available token telemetry is never exercised**

All successful action and grader receipts set `tokenUsage` to `unavailable`, and the only object-shaped token fixture is deliberately invalid, so propagation and aggregation of valid input, output, and total token counts have no observable-contract coverage.

**Fix:** Add a successful fixture with valid numeric token usage for both variants and assert the emitted per-case and overall token metrics and deltas.

**Review by howarewoo — COMMENTED** [review]
Review summary: no blocking findings; four nits identify focused regression-coverage gaps.

**Status: APPROVED** — No blocking findings, 4 nit(s). See inline comments.

Audited by woostack-review · Host: omp · Provider: openai-codex · Model: gpt-5.6-sol

File: skills/woostack-eval/scripts/aggregate/core.mjs (hunk 8 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/core.mjs:513]: Added focused evidence and output containment rejection cases. Each asserts the command fails with the containment error and publishes no output. Fixed in 871260f.

File: skills/woostack-eval/scripts/aggregate/safe-access.mjs (hunk 9 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/aggregate/safe-access.mjs:488]: Added mode-0755 and symlink publication-parent rejection cases, including checks that neither target contains a final nor temporary file. Fixed in 871260f.

File: skills/woostack-eval/scripts/tests/test-aggregate.sh (hunk 10 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/tests/test-aggregate.sh:43]: Added a complete action and grader fixture using sessionIdentity with model null. Fixture self-validation checks the identity contract and aggregation succeeds. Fixed in 871260f.

File: skills/woostack-eval/scripts/tests/test-aggregate.sh (hunk 11 above)
Comment (howarewoo) [reply skills/woostack-eval/scripts/tests/test-aggregate.sh:185]: Added successful concrete token telemetry for both variants and assertions for per-result values, aggregate means, unavailable variances, and deltas. Fixed in 871260f.
